_posts/2020-09-29-introducing-torch-for-r/introducing_torch_for_R.Rmd

```diff
@@ -0,0 +1,424 @@
+
+Here, a bit of calculation is necessary. We go through all actions that happen in `forward()` -- if they change shapes, we keep track of the transformation; if they don't, we ignore them.
+
+So, we start with input tensors of shape `batch_size x 1 x 28 x 28`. Then,
+
+-   `nn_conv2d(1, 32, 3)` , or equivalently, `nn_conv2d(in_channels = 1, out_channels = 32, kernel_size = 3),`applies a convolution with kernel size 3, stride 1 (the default) and no padding (the default). We can consult the [documentation](https://mlverse.github.io/torch/reference/nn_conv2d.html) to look up the resulting output size, or just intuitively reason that with a kernel of size 3 and no padding, the image will shrink by one pixel in each direction, resulting in a spatial resolution of `26 x 26`. *Per channel*, that is. Thus, the actual output shape is `batch_size x 32 x 26 x 26` . Next,
+
+-   `nnf_relu()` applies ReLU activation, in no way touching the shape. Next is
+
+-   `nn_conv2d(32, 64, 3)`, another convolution with zero padding and kernel size 3. Output size now is `batch_size x 64 x 24 x 24` . Now, the second
+
+-   `nnf_relu()` again does nothing to the output shape, but
+
+-   `nnf_max_pool2d(2)` (equivalently: `nnf_max_pool2d(kernel_size = 2)`) does: It applies max pooling over regions of extension `2 x 2`, thus downsizing the output to a format of `batch_size x 32 x 12 x 12` . Now,
+
+-   `nn_dropout2d(0.25)` is a no-op, shape-wise, but if we want to apply a linear layer later, we need to merge all of the *channels*, *height* and *width* axes into a single dimension. This is done in
+
+-   `torch_flatten(start_dim = 2)`. Output shape is now `batch_size * 9216` , since `64 * 12 * 12 = 9216` . Thus here we have the `9216` input connections fed into the
+
+-   `nn_linear(9216, 128)` discussed above. Again,
+
+-   `nnf_relu()` and `nn_dropout2d(0.5)` leave dimensions as they are, and finally,
+
+-   `nn_linear(128, 10)` gives us the desired output scores, one for each of the ten classes.
+
+Now you'll be thinking, -- what if my network is more complicated? Calculations could become pretty cumbersome. Luckily, with `torch`'s flexibility, there is another way. Since every layer is callable *in isolation*, we can just ... create some sample data and see what happens!
+
+Here is a sample "image" -- or more precisely, a one-item batch containing it:
+
+```{r}
+x <- torch_randn(c(1, 1, 28, 28))
+```
+
+What if we call the first *conv2d* module on it?
+
+```{r}
+conv1 <- nn_conv2d(1, 32, 3)
+conv1(x)$size()
+```
+
+    [1]  1 32 26 26
+
+Or both *conv2d* modules?
+
+```{r}
+conv2 <- nn_conv2d(32, 64, 3)
+(conv1(x) %>% conv2())$size()
+```
+
+    [1]  1 64 24 24
+
+And so on. This is just one example illustrating how `torch`s flexibility makes developing neural nets easier.
+
+Back to the main thread. We instantiate the model, and we ask `torch` to allocate its weights (parameters) on the GPU:
+
+```{r}
+model <- net()$to(device = "cuda")
```

Comment:
I would split this into 2 lines. One to initialize the model and the other to move to the GPU.

Reply:
makes sense